schemas/dns.yml

```diff
@@ -0,0 +1,254 @@
+        This value is set by the originator of a query and copied into the
+        response.
+      example: QUERY
+
+    - name: flags.authoritative
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying that the responding server is an authority for
+        the domain name used in the question.
+
+    - name: flags.recursion_available
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying whether recursive query support is available in the
+        name server.
+
+    - name: flags.recursion_desired
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying that the client directs the server to pursue a
+        query recursively. Recursive query support is optional.
+
+    - name: flags.authentic_data
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying that the recursive server considers the response
+        authentic.
+
+    - name: flags.checking_disabled
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying that the client disables the server
+        signature validation of the query.
+
+    - name: flags.truncated_response
+      level: extended
+      type: boolean
+      description: >
+        A DNS flag specifying that only the first 512 bytes of the reply were
+        returned.
+
+    - name: response_code
+      level: extended
+      type: keyword
+      description: The DNS status code.
+      example: NOERROR
+
+    - name: question.name
+      level: core
+      type: keyword
+      description: >
+        The domain name being queried. If the name field contains non-printable
+        characters (below 32 or above 126), then those characters are represented
+        as escaped base 10 integers (\DDD). Back slashes and quotes are escaped.
+        Tabs, carriage returns, and line feeds are converted to \t, \r, and
+        \n respectively.
+      example: www.google.com.
+
+    - name: question.type
+      level: core
+      type: keyword
+      description: The type of records being queried.
+      example: AAAA
+
+    - name: question.class
+      level: core
+      type: keyword
+      description: The class of of records being queried.
+      example: IN
+
+    - name: question.etld_plus_one
+      level: core
+      type: keyword
+      description: The effective top-level domain (eTLD) plus one more label.
+        For example, the eTLD+1 for "foo.bar.golang.org." is "golang.org.".
+        The data for determining the eTLD comes from an embedded copy of the
+        data from http://publicsuffix.org.
+      example: amazon.co.uk.
+
+    - name: question.size
+      level: extended
+      type: keyword
+      description: The size of name being queried (in bytes).
+
+    - name: answers
+      level: extended
+      type: object
+      description: >
```

Comment:
This is the array of objects I'm worried about.

So first, what do they answers look like? Here's an example of an A record query for "collector.githubapp.com":

```JSON
# dns.answers contains:
[
{
  "data": "analytics-collector-28944298.us-east-1.elb.amazonaws.com",
  "name": "collector.githubapp.com",
  "type": "CNAME",
  "class": "IN",
  "ttl": "738"
},
{
  "type": "A",
  "class": "IN",
  "ttl": "21",
  "data": "34.225.79.123",
  "name": "analytics-collector-28944298.us-east-1.elb.amazonaws.com"
},
{
  "class": "IN",
  "ttl": "21",
  "data": "52.205.37.215",
  "name": "analytics-collector-28944298.us-east-1.elb.amazonaws.com",
  "type": "A"
},
{
  "type": "A",
  "class": "IN",
  "ttl": "21",
  "data": "52.205.238.88",
  "name": "analytics-collector-28944298.us-east-1.elb.amazonaws.com"
}
]
```

As for "is it searchable?" Well Elasticsearch will return the document if you search for `dns.answers.name:analytics-collector-28944298.us-east-1.elb.amazonaws.com`. 

Although leveraging documents like this in Kibana is pretty difficult.

When looking for the same exchange on the Suricata side, I searched for `suricata.eve.dns.rrname:collector.githubapp.com` I only got the A query and one CNAME answer, but none of the A record answers. By selecting the flow ID and removing the filter for the domain search query, then I saw them all (A request, CNAME answer, A answers), like we have above for Packetbeat.

So this multi-step process is not ideal either. If we decide to denormalize to have one document per answer, I wonder if we could get to the whole list of results only by querying for the initial DNS query, and have the fields structured such that you get all documents at once.

Reply:
+1 on having separate events instead of nested docs.

Reply:
After much discussion with the SIEM team, we've decided to keep the array of object, because it was the most straightforward to query and make sense of.

To simplify the consumption of the resolved IPs -- which is often the goal of looking at DNS -- we're adding field `dns.resolved_ip` which is meant to contain an array of all IPs. It's type is `ip`.

Reply:
I'm still concerned about introducing nested objects in ECS. Is Kibana supporting this properly today, meaning can users also query it without sending queries directly to Elasticsearch?

At the same time, do not want to block this PR on this. Perhaps we can remove it from the PR itself and open a separate one to get this in and discuss further?

Reply:
No, Kibana doesn't support arrays of objects better today. I agree that an array of objects is difficult to use in Kibana dashboards, but it's still the best option we have here, after considering others.

**1 vs multiple events per DNS exchange**

It's been discussed whether splitting the exchanges in multiple events would help model this better (1 event for the query, 1 event for each of the answers). But what we found out is while that would avoid having an array of object, it would be more difficult to query. It means consumers have to do an in memory join (more than one API call), to get their full DNS exchanges. This is merely annoying in dashboards, but can actually be a deal breaker for queries in ML or watcher, where you can't specify back to back queries to perform that join.

Suricata used to be the only one of the 3 that separated DNS answers like this, and both Zeek and Packetbeat followed the one full exchange event. But starting with Suricata 4.1, they're moving to one DNS exchange per document (for this you need to set `dns.version:2` Suricata's yaml config).

With this, we have all 3 sources going to one event per full DNS exchange. Once we settle on this approach, there's the question of how do we represent them?

**How to represent the exchange in one document**

The options here are either to have multiple arrays of simple values, or one array of objects, where each object contains the relevant keys.

Zeek takes a simple approach to this, and uses multiple arrays of values, where their position in the array is correlated. But Zeek only uses 2 fields: answers and TTLs, and is therefore discarding a lot of information. To more fully model DNS exchanges, we need to define up to 5 such arrays (name, type, answer, ttl, class). But pushing a little further and looking at DNSSEC, we would potentially need 15 fields to describe the exchange.

While I don't think we need to model all DNSSEC fields just yet, considering this extreme case made the multiple arrays solution feel very hacky.

**The compromise**

What we settled on is that we keep the `dns.answers` array of object, but we also extract all IPs seen in there to `dns.resolved_ip`, which is of datatype `ip`. What users are often looking for with DNS is the IP addresses, so having all of them extracted out to this array of IPs makes them easy to query, easy enough to visualize, and we think is a good compromise. The array of objects in this situation is no longer the main source of information, but is a record of all details of the exchange, available for close inspection.

Reply:
> can users also query it without sending queries directly to Elasticsearch?

Where we have an array of objects, like DNS'

```
dns.answers: [
  { "name": "example.com", "data": "ww1.example.com", "ttl": 42 ...},
  { "name": "ww1.example.com", "data": "10.10.10.10", "ttl": 42 ...}
]
```

It's important to understand that

1. The field definitions for nested fields such as `dns.answers.ttl` (datatype `long`) are respected by Elasticsearch.
2. The searches/filters on any of these fields will work just as well as if there was no array of objects.
3. Aggregations on these fields will also work as expected. If one domain has 4 IP addresses in its A record for a domain (and therefore 4 objects with `data` set to an IP), all 4 IPs will be counted in the aggs.

So the problems with these fields in Kibana is that you can only visualize them as a JSON blob, and therefore you can't use the mouse to add filters. But if the main information piece of information you're looking for is elsewhere (in `dns.resolved_ip`), then this isn't as much a problem.